elasticsearch-akka/src/main/scala/com/sumologic/elasticsearch/akkahelpers/ScanAndScrollSourceGraph.scala

```diff
@@ -0,0 +1,87 @@
+/**
+  * Licensed to the Apache Software Foundation (ASF) under one
+  * or more contributor license agreements.  See the NOTICE file
+  * distributed with this work for additional information
+  * regarding copyright ownership.  The ASF licenses this file
+  * to you under the Apache License, Version 2.0 (the
+  * "License"); you may not use this file except in compliance
+  * with the License.  You may obtain a copy of the License at
+  *
+  * http://www.apache.org/licenses/LICENSE-2.0
+  *
+  * Unless required by applicable law or agreed to in writing,
+  * software distributed under the License is distributed on an
+  * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+  * KIND, either express or implied.  See the License for the
+  * specific language governing permissions and limitations
+  * under the License.
+  */
+package com.sumologic.elasticsearch.akkahelpers
+
+import akka.NotUsed
+import akka.stream._
+import akka.stream.scaladsl.{Flow, GraphDSL, Merge, Source, Unzip}
+import akka.stream.stage.{GraphStage, GraphStageLogic, InHandler, OutHandler}
+import com.sumologic.elasticsearch.restlastic.RestlasticSearchClient.ReturnTypes.{ScrollId, SearchResponse}
+import com.sumologic.elasticsearch.restlastic.ScrollClient
+import com.sumologic.elasticsearch.restlastic.dsl.Dsl._
+
+import scala.concurrent.ExecutionContext
+
+object ScanAndScrollSourceGraph {
```

Comment:
This can be done more succinctly, using the Akka-streams `Source` DSL method [`Source.unfoldAsync`](https://doc.akka.io/api/akka/current/akka/stream/scaladsl/Source$.html#unfoldAsync[S,E](s:S)(f:S=%3Escala.concurrent.Future[Option[(S,E)]]):akka.stream.scaladsl.Source[E,akka.NotUsed]) instead of hand-creating a `GraphStage`.

Here's about how you could go about it:

Given are 2 methods from `ScrollClient`, one that starts a scroll, and another that continues the scroll:
(eliding other parameters here...)
```
def start(): Future[(ScrollId, SearchResponse)]
def next(id: ScrollId): Future[(ScrollId, SearchResponse)]
```

Before we can unfold, we need to do 2 things:
1. The source created by`unfoldAsync` terminates when the underlying method returns `None`. Here, we want termination when search response is empty. So,
    ```
    def emptyAsNone[A](resp: (A, SearchResponse)): Option[(A, SearchResponse)] = {
      if (resp._2.length == 0) None
      else Some(resp)
    }
    ```
2. For purposes of passing to `unfoldAsync`, we need to combine the `start` and `next` into a single method, which takes inputs and produces output in terms of `StartOrScrollId` instead of `ScrollId`: 
    ```
    case object Start
    type StartOrScrollId = Either[Start.type, ScrollId]

    def withStartOrScrollId[T](resp: (ScrollId, T)): (StartOrScrollId, T) = {
      (Right(resp._1), resp._2)
    }

    def startOrNext(id: StartOrScrollId)
    : Future[Option[(StartOrScrollId, SearchResponse)]] = {
      id match {
        case Left(_)         => start()       .map (obj => emptyAsNone(withStartOrScrollId(obj)))
        case Right(scrollId) => next(scrollId).map (obj => emptyAsNone(withStartOrScrollId(obj)))
      }
    }
    ```

Finally, all that remains is unfolding to produce a `Source`:
```
val scanAndScrollSource: Source[SearchResponse, Unit] = {
  Source.unfoldAsync[StartOrScrollId, SearchResponse](Left(Start))(startOrNext) 
}
```

Reply:
@kumar-avijit whoa this is pretty cool. How would you go about detecting the end of the stream though?

Reply:
The `unfold`ing stops when the method being unfolded returns a `None`:
```
def unfoldAsync[S, E](s: S)(f: (S) ⇒ Future[Option[(S, E)]]): Source[E, NotUsed]
```
Btw, I learned today that the Dsl provides a few other variants such as [`unfoldResource`](https://doc.akka.io/api/akka/current/akka/stream/scaladsl/Source$.html#unfoldResource[T,S](create:()=%3ES,read:S=%3EOption[T],close:S=%3EUnit):akka.stream.scaladsl.Source[T,akka.NotUsed]) and [`unfoldResourceAsync`](https://doc.akka.io/api/akka/current/akka/stream/scaladsl/Source$.html#unfoldResourceAsync[T,S](create:()=%3Escala.concurrent.Future[S],read:S=%3Escala.concurrent.Future[Option[T]],close:S=%3Escala.concurrent.Future[akka.Done]):akka.stream.scaladsl.Source[T,akka.NotUsed]) which can come in handy if you want to add pre- and post- stream methods, such as when you want to initialize or close a resource at the start or end of the stream. Just fyi, in case you find yourself needing to do that.